### PR TITLE
amp and Zero bf16 support

### PR DESCRIPTION
**Pull request by ht-zhou**
Support bf16 data type in amp and zero.

### CODE REVIEW

**Comment by FrankLeeeee** [comment]
Hi @ht-zhou , this PR has some conflicts with the main branch, can you fix them first?

File: colossalai/amp/__init__.py [hunk 1]
```diff
@@ -13,14 +13,15 @@
 __all__ = ['convert_to_amp', 'convert_to_naive_amp', 'convert_to_apex_amp', 'convert_to_torch_amp', 'AMP_TYPE']
 
 
-def convert_to_amp(model: nn.Module, optimizer: Optimizer, criterion: _Loss, mode: AMP_TYPE, amp_config: Config = None):
+def convert_to_amp(model: nn.Module, optimizer: Optimizer, criterion: _Loss, mode: AMP_TYPE, use_bf16: bool = False, amp_config: Config = None):
```
Comment (1SAA) [inline colossalai/amp/__init__.py:16]: passing a bool paramter `use_bf16` is not a good way. could you passing a precision type? you can use `.to(torch.bfloat16)` or `.to(torch.float16)` to control your target precision type.

File: colossalai/kernel/cuda_native/csrc/cpu_adam.cpp [hunk 2]
```diff
@@ -446,8 +446,8 @@ void Adam_Optimizer::step(size_t step, float lr, float beta1, float beta2,
   this->IncrementStep(step, beta1, beta2);
   this->update_state(lr, epsilon, weight_decay, bias_correction);
   this->Step_8(params_ptr, grads_ptr, exp_avg_ptr, exp_avg_sq_ptr,
-               params_c.numel(), (params.options().dtype() == at::kHalf),
-               (grads.options().dtype() == at::kHalf), loss_scale);
```
Comment (1SAA) [inline colossalai/kernel/cuda_native/csrc/cpu_adam.cpp:450]: why do you delete paramter `loss_scale`?

File: colossalai/testing/comparison.py [hunk 3]
```diff
@@ -13,7 +13,7 @@ def assert_not_equal(a: Tensor, b: Tensor):
 
 
 def assert_close(a: Tensor, b: Tensor, rtol: float = 1e-5, atol: float = 1e-8):
-    assert torch.allclose(a, b, rtol=rtol, atol=atol), f'expected a and b to be close but they are not, {a} vs {b}'
+    assert torch.allclose(a.cuda().float(), b.cuda().float(), rtol=rtol, atol=atol), f'expected a and b to be close but they are not, {a} vs {b}'
```
Comment (1SAA) [inline colossalai/testing/comparison.py:16]: why you move all tensors to cuda? can you add some parameters in this function to control this?

**Comment by binmakeswell** [comment]
Hi @ht-zhou Is this PR still active? There are many conflicts.

**Pull request closed without merging**